File: backend/api/freebill.py
from __future__ import annotations

from typing import Any, Literal

from fastapi import APIRouter, Depends, File, HTTPException, Query, UploadFile
from pydantic import BaseModel, Field
from sqlmodel import Session

from backend.core.auth import get_current_active_user
from backend.core.feature_access_guard import require_feature_access_dependency
from backend.core.freebill import (
    clear_freebill_record_overrides,
    get_freebill_dashboard,
    get_freebill_status,
    import_alipay_csv_bytes,
    import_ccb_excel_bytes,
    import_wechat_excel_bytes,
    list_freebill_interpret_rules,
    list_freebill_category_branch_records,
    list_freebill_filter_options,
    list_freebill_records,
    recompute_freebill_interpretation,
    save_freebill_interpret_rules,
    upsert_freebill_category_branch_overrides,
    upsert_freebill_category_branch_manual_overrides,
    upsert_freebill_record_manual_overrides,
    upsert_freebill_record_overrides,
)
from backend.core.freebill_sheet import get_freebill_sheet_workbook, refresh_freebill_sheet_workbook
from backend.db import get_session
from backend.models import User


router = APIRouter(
    dependencies=[Depends(require_feature_access_dependency("notes.freebill"))],
)


FreebillProgramRuleAction = Literal["include", "exclude", "filter"]
FreebillProgramMatcherKind = Literal["all", "none", "field", "full_text_contains"]
FreebillInterpretRuleMatcherKind = Literal["all", "none", "field", "full_text_contains"]
FreebillProgramOperator = Literal[
    "eq",
    "neq",
    "in",
    "not_in",
    "contains",
    "not_contains",
    "gte",
    "lte",
    "between",
    "year",
]
FreebillInterpretRuleOperator = Literal[
    "eq",
    "neq",
    "in",
    "not_in",
    "contains",
    "not_contains",
    "gt",
    "gte",
    "lt",
    "lte",
]
FreebillCategoryDimension = Literal[
    "standard_direction",
    "standard_nature",
    "type",
    "counterparty",
]


class FreebillProgramMatcher(BaseModel):
    kind: FreebillProgramMatcherKind = "all"
    field: str | None = None
    op: FreebillProgramOperator | None = None
    value: Any = None
    values: list[Any] = Field(default_factory=list)
    ignore_case: bool = True


class FreebillProgramRule(BaseModel):
    action: FreebillProgramRuleAction = "include"
    matcher: FreebillProgramMatcher = Field(default_factory=FreebillProgramMatcher)


class FreebillProgramChannel(BaseModel):
    default: bool = False
    rules: list[FreebillProgramRule] = Field(default_factory=list)


class FreebillInterpretRuleMatcher(BaseModel):
    kind: FreebillInterpretRuleMatcherKind = "field"
    field: str | None = "product_name"
    op: FreebillInterpretRuleOperator | None = "contains"
    value: Any = None
    values: list[Any] = Field(default_factory=list)
    ignore_case: bool = True


class FreebillInterpretRuleItem(BaseModel):
    id: int | None = None
    name: str = ""
    enabled: bool = True
    order_index: int = 0
    matcher: FreebillInterpretRuleMatcher = Field(default_factory=FreebillInterpretRuleMatcher)
    set_direction: Literal["支出", "收支", "收入"] | None = None
    set_nature: Literal["常规", "借贷", "理财", "转账", "流水"] | None = None
    note: str | None = None


class FreebillInterpretRuleSettings(BaseModel):
    signed_category_values: bool = False
    built_in_rules: dict[str, bool] = Field(default_factory=dict)


class FreebillInterpretRulesRequest(BaseModel):
    rules: list[FreebillInterpretRuleItem] = Field(default_factory=list)
    settings: FreebillInterpretRuleSettings = Field(default_factory=FreebillInterpretRuleSettings)


class FreebillDashboardProgramRequest(BaseModel):
    program: FreebillProgramChannel = Field(default_factory=FreebillProgramChannel)
    programs: list[FreebillProgramChannel] = Field(default_factory=list)
    trend_granularity: Literal["day", "week", "month", "year"] = "month"
    trend_standard_nature: Literal["常规", "借贷", "理财", "转账", "流水"] | None = None
    category_dimensions: list[FreebillCategoryDimension] = Field(default_factory=list)


class FreebillCategoryPathItem(BaseModel):
    dimension: FreebillCategoryDimension
    value: str


class FreebillCategoryBranchRecordsRequest(BaseModel):
    program: FreebillProgramChannel = Field(default_factory=FreebillProgramChannel)
    programs: list[FreebillProgramChannel] = Field(default_factory=list)
    path: list[FreebillCategoryPathItem] = Field(default_factory=list)
    direction: str | None = None
    category: str | None = None
    counterparty: str | None = None
    limit: int = Field(default=10, ge=1, le=50)
    offset: int = Field(default=0, ge=0)
    sort_by: Literal["amount", "create_time", "source", "product_name", "remark"] = "amount"
    sort_order: Literal["asc", "desc"] = "desc"


class FreebillRecordOverrideRequest(BaseModel):
    trade_nos: list[str] = Field(min_length=1)
    direction: str = "不计收支"
    category: str = "流水"
    note: str | None = None


class FreebillCategoryBranchOverrideRequest(FreebillCategoryBranchRecordsRequest):
    override_direction: str = "不计收支"
    override_category: str = "流水"
    note: str | None = None


class FreebillRecordOverrideClearRequest(BaseModel):
    trade_nos: list[str] = Field(min_length=1)


class FreebillRecordManualOverrideRequest(BaseModel):
    trade_no: str
    overrides: dict[str, object] = Field(default_factory=dict)
    note: str | None = None


class FreebillCategoryBranchManualOverrideRequest(FreebillCategoryBranchRecordsRequest):
    overrides: dict[str, object] = Field(default_factory=dict)
    note: str | None = None


@router.get("/status")
def get_status():
    return get_freebill_status()


@router.get("/dashboard")
def get_dashboard(
    start_date: str | None = Query(default=None, pattern=r"^\d{4}-\d{2}-\d{2}$"),
    end_date: str | None = Query(default=None, pattern=r"^\d{4}-\d{2}-\d{2}$"),
    source: str | None = Query(default=None),
    direction: str | None = Query(default=None),
    category: str | None = Query(default=None),
    q: str | None = Query(default=None),
    trend_granularity: Literal["day", "week", "month", "year"] = Query(default="month"),
    trend_standard_nature: Literal["常规", "借贷", "理财", "转账", "流水"] | None = Query(default=None),
):
    return get_freebill_dashboard(
        start_date=start_date,
        end_date=end_date,
        source=source,
        direction=direction,
        category=category,
        q=q,
        trend_granularity=trend_granularity,
        trend_standard_nature=trend_standard_nature,
    )


@router.post("/dashboard-program")
def get_dashboard_by_program(payload: FreebillDashboardProgramRequest):
    try:
        return get_freebill_dashboard(
            program=payload.program.model_dump(),
            programs=[program.model_dump() for program in payload.programs] or None,
            trend_granularity=payload.trend_granularity,
            trend_standard_nature=payload.trend_standard_nature,
            category_dimensions=payload.category_dimensions or None,
        )
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc


@router.get("/records")
def get_records(
    start_date: str | None = Query(default=None, pattern=r"^\d{4}-\d{2}-\d{2}$"),
    end_date: str | None = Query(default=None, pattern=r"^\d{4}-\d{2}-\d{2}$"),
    source: str | None = Query(default=None),
    direction: str | None = Query(default=None),
    category: str | None = Query(default=None),
    q: str | None = Query(default=None),
    limit: int = Query(default=80, ge=1, le=500),
    offset: int = Query(default=0, ge=0),
):
    return list_freebill_records(
        start_date=start_date,
        end_date=end_date,
        source=source,
        direction=direction,
        category=category,
        q=q,
        limit=limit,
        offset=offset,
    )


@router.post("/category-branch-records")
def get_category_branch_records(payload: FreebillCategoryBranchRecordsRequest):
    try:
        return list_freebill_category_branch_records(
            program=payload.program.model_dump(),
            programs=[program.model_dump() for program in payload.programs] or None,
            path=[item.model_dump() for item in payload.path] or None,
            direction=payload.direction,
            category=payload.category,
            counterparty=payload.counterparty,
            limit=payload.limit,
            offset=payload.offset,
            sort_by=payload.sort_by,
            sort_order=payload.sort_order,
        )
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc


@router.get("/filter-options")
def get_filter_options():
    return list_freebill_filter_options()


@router.get("/interpret-rules")
def get_interpret_rules():
    return list_freebill_interpret_rules()


@router.put("/interpret-rules")
def save_interpret_rules(payload: FreebillInterpretRulesRequest):
    try:
        return save_freebill_interpret_rules(
            [item.model_dump() for item in payload.rules],
            settings=payload.settings.model_dump(),
        )
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc


@router.post("/interpret-rules/recompute")
def recompute_interpret_rules():
    return recompute_freebill_interpretation()


@router.post("/record-overrides")
def apply_record_overrides(payload: FreebillRecordOverrideRequest):
    try:
        return upsert_freebill_record_overrides(
            payload.trade_nos,
            direction=payload.direction,
            category=payload.category,
            note=payload.note,
        )
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc


@router.put("/record-manual-overrides")
def apply_record_manual_overrides(payload: FreebillRecordManualOverrideRequest):
    try:
        return upsert_freebill_record_manual_overrides(
            payload.trade_no,
            payload.overrides,
            note=payload.note,
        )
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc


@router.post("/category-branch-overrides")
def apply_category_branch_overrides(payload: FreebillCategoryBranchOverrideRequest):
    try:
        return upsert_freebill_category_branch_overrides(
            program=payload.program.model_dump(),
            programs=[program.model_dump() for program in payload.programs] or None,
            path=[item.model_dump() for item in payload.path] or None,
            direction=payload.direction,
            category=payload.category,
            counterparty=payload.counterparty,
            override_direction=payload.override_direction,
            override_category=payload.override_category,
            note=payload.note,
        )
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc


@router.put("/category-branch-manual-overrides")
def apply_category_branch_manual_overrides(payload: FreebillCategoryBranchManualOverrideRequest):
    try:
        return upsert_freebill_category_branch_manual_overrides(
            program=payload.program.model_dump(),
            programs=[program.model_dump() for program in payload.programs] or None,
            path=[item.model_dump() for item in payload.path] or None,
            overrides=payload.overrides,
            note=payload.note,
        )
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc


@router.post("/record-overrides/clear")
def clear_record_overrides(payload: FreebillRecordOverrideClearRequest):
    try:
        return clear_freebill_record_overrides(payload.trade_nos)
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc


@router.get("/sheet-workbook")
def get_freebill_sheet_file(
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_active_user),
):
    return get_freebill_sheet_workbook(session, user_id=int(current_user.id))


@router.post("/sheet-workbook/refresh")
def refresh_freebill_sheet_file(
    session: Session = Depends(get_session),
    current_user: User = Depends(get_current_active_user),
):
    return refresh_freebill_sheet_workbook(
        session,
        user_id=int(current_user.id),
        actor_user_id=int(current_user.id),
    )


@router.post("/import/{source}")
async def import_files(
    source: Literal["alipay", "wechat", "ccb"],
    files: list[UploadFile] = File(...),
    header_row: int = Query(default=24, ge=0, le=200),
):
    if not files:
        raise HTTPException(status_code=400, detail="请上传账单文件")

    results: list[dict] = []
    for upload_file in files:
        filename = upload_file.filename or "账单文件"
        try:
            content = await upload_file.read()
            if not content:
                raise ValueError("文件内容为空")
            if source == "alipay":
                result = import_alipay_csv_bytes(filename, content, header_row=header_row)
            elif source == "wechat":
                result = import_wechat_excel_bytes(filename, content)
            else:
                result = import_ccb_excel_bytes(filename, content)
            results.append({"status": "success", **result})
        except Exception as exc:
            results.append(
                {
                    "status": "error",
                    "filename": filename,
                    "processed": 0,
                    "inserted": 0,
                    "skipped": 0,
                    "error": str(exc),
                }
            )
        finally:
            await upload_file.close()

    return {
        "source": source,
        "results": results,
        "processed": sum(int(item.get("processed") or 0) for item in results),
        "inserted": sum(int(item.get("inserted") or 0) for item in results),
        "skipped": sum(int(item.get("skipped") or 0) for item in results),
        "error_count": sum(1 for item in results if item.get("status") == "error"),
    }

File: frontend/src/api/freebill.ts
import api from '@/api'

export type FreebillImportSource = 'alipay' | 'wechat' | 'ccb'
export type FreebillTrendGranularity = 'day' | 'week' | 'month' | 'year'
export type FreebillStandardNature = '常规' | '借贷' | '理财' | '转账' | '流水'
export type FreebillStandardDirection = '支出' | '收支' | '收入'
export type FreebillCategoryBranchSortBy = 'amount' | 'create_time' | 'source' | 'product_name' | 'remark'
export type FreebillSortOrder = 'asc' | 'desc'
export type FreebillCategoryDimension = 'standard_direction' | 'standard_nature' | 'type' | 'counterparty'
export type FreebillProgramMatcherKind = 'all' | 'none' | 'field' | 'full_text_contains'
export type FreebillInterpretRuleMatcherKind = 'all' | 'none' | 'field' | 'full_text_contains'
export type FreebillProgramRuleAction = 'include' | 'exclude' | 'filter'
export type FreebillProgramOperator =
  | 'eq'
  | 'neq'
  | 'in'
  | 'not_in'
  | 'contains'
  | 'not_contains'
  | 'gte'
  | 'lte'
  | 'between'
  | 'year'
export type FreebillInterpretRuleOperator =
  | 'eq'
  | 'neq'
  | 'in'
  | 'not_in'
  | 'contains'
  | 'not_contains'
  | 'gt'
  | 'gte'
  | 'lt'
  | 'lte'

export interface FreebillStatus {
  exists: boolean
  work_dir: string
  db_path: string
  total_records: number
  min_date: string | null
  max_date: string | null
  last_imported_at: number | null
  raw_file_count: number
}

export interface FreebillSummary {
  total_income: number
  total_expense: number
  total_ignore: number
  total_other: number
  total_count: number
  balance: number
}

export interface FreebillSourceBreakdown {
  source: string
  count: number
  income: number
  expense: number
}

export interface FreebillCategoryStat {
  name: string
  value: number
  count: number
  group_count?: number
  is_remainder?: boolean
  remainder_items?: FreebillCategoryStat[]
  dimension?: FreebillCategoryDimension | null
  path?: FreebillCategoryPathItem[]
  children?: FreebillCategoryStat[]
}

export interface FreebillCategoryPathItem {
  dimension: FreebillCategoryDimension
  value: string
}

export interface FreebillMonthlyTrendItem {
  month: string
  income: number
  expense: number
  other?: number
  income_count: number
  expense_count: number
  ignore_count?: number
  other_count?: number
  count: number
}

export interface FreebillDashboard {
  summary: FreebillSummary
  sources: FreebillSourceBreakdown[]
  expense_categories: FreebillCategoryStat[]
  income_categories: FreebillCategoryStat[]
  category_tree?: FreebillCategoryStat[]
  category_dimensions?: FreebillCategoryDimension[]
  monthly_trend: FreebillMonthlyTrendItem[]
  trend_granularity?: FreebillTrendGranularity
}

export interface FreebillRecord {
  id: number
  source: string | null
  trade_no: string | null
  merchant_order_no: string | null
  create_time: string | null
  pay_time: string | null
  modify_time: string | null
  location: string | null
  type: string | null
  counterparty: string | null
  product_name: string | null
  amount: number | null
  direction: string | null
  status: string | null
  service_fee: number | null
  refund_amount: number | null
  remark: string | null
  fund_status: string | null
  account_no: string | null
  currency: string | null
  cash_type: string | null
  account_balance: number | null
  raw_sequence: string | null
  standard_nature?: string | null
  standard_direction?: string | null
  raw_direction?: string | null
  raw_type?: string | null
  raw_values?: Record<string, unknown>
  manual_overrides?: Record<string, unknown>
  imported_at: number | null
  has_record_override?: boolean | number | null
}

export interface FreebillRecordPage {
  total: number
  items: FreebillRecord[]
}

export interface FreebillFilterOptions {
  sources: string[]
  directions: string[]
  types: string[]
  categories: string[]
}

export interface FreebillInterpretRuleField {
  value: string
  label: string
  mode: 'text' | 'number'
}

export interface FreebillInterpretRuleOperatorOption {
  value: FreebillInterpretRuleOperator
  label: string
}

export interface FreebillBuiltInInterpretRule {
  key: string
  name: string
  enabled: boolean
  target_nature?: FreebillStandardNature | null
  matcher_text: string
  result_text: string
  note?: string | null
}

export interface FreebillInterpretRuleMatcher {
  kind: FreebillInterpretRuleMatcherKind
  field?: string | null
  op?: FreebillInterpretRuleOperator | null
  value?: unknown
  values?: unknown[]
  ignore_case?: boolean
}

export interface FreebillInterpretRule {
  id?: number | null
  name: string
  enabled: boolean
  order_index: number
  matcher: FreebillInterpretRuleMatcher
  set_direction?: FreebillStandardDirection | null
  set_nature?: FreebillStandardNature | null
  note?: string | null
  match_count?: number
}

export interface FreebillInterpretRuleSettings {
  signed_category_values: boolean
  built_in_rules: Record<string, boolean>
}

export interface FreebillInterpretRulesPayload {
  settings: FreebillInterpretRuleSettings
  fields: FreebillInterpretRuleField[]
  operators: FreebillInterpretRuleOperatorOption[]
  directions: FreebillStandardDirection[]
  natures: FreebillStandardNature[]
  built_in_rules: FreebillBuiltInInterpretRule[]
  rules: FreebillInterpretRule[]
}

export interface FreebillInterpretRecomputeResult {
  total: number
  updated: number
  rules: FreebillInterpretRule[]
  recomputed_at: number
}

export interface FreebillQueryParams {
  start_date?: string
  end_date?: string
  source?: string
  direction?: string
  category?: string
  q?: string
  trend_granularity?: FreebillTrendGranularity
  limit?: number
  offset?: number
}

export interface FreebillProgramMatcher {
  kind: FreebillProgramMatcherKind
  field?: string | null
  op?: FreebillProgramOperator | null
  value?: unknown
  values?: unknown[]
  ignore_case?: boolean
}

export interface FreebillProgramRule {
  action: FreebillProgramRuleAction
  matcher: FreebillProgramMatcher
}

export interface FreebillProgramChannel {
  default: boolean
  rules: FreebillProgramRule[]
}

export interface FreebillDashboardProgramRequest {
  program: FreebillProgramChannel
  programs?: FreebillProgramChannel[]
  trend_granularity?: FreebillTrendGranularity
  trend_standard_nature?: FreebillStandardNature | null
  category_dimensions?: FreebillCategoryDimension[]
}

export interface FreebillCategoryBranchRecordsRequest {
  program: FreebillProgramChannel
  programs?: FreebillProgramChannel[]
  path?: FreebillCategoryPathItem[]
  direction?: string | null
  category?: string | null
  counterparty?: string | null
  limit?: number
  offset?: number
  sort_by?: FreebillCategoryBranchSortBy
  sort_order?: FreebillSortOrder
}

export interface FreebillImportResultItem {
  status: 'success' | 'error'
  filename: string
  processed: number
  inserted: number
  updated: number
  skipped: number
  error?: string
}

export interface FreebillImportResult {
  source: FreebillImportSource
  results: FreebillImportResultItem[]
  processed: number
  inserted: number
  updated: number
  skipped: number
  error_count: number
}

export interface FreebillWechatLocalSyncResult {
  filename: string
  format: string
  processed: number
  inserted: number
  updated: number
  skipped: number
  scanned: number
  parsed: number
  ignored: number
  source_modified_at: number
  source_sync_error?: string | null
  payment: {
    scanned: number
    parsed: number
    ignored: number
    reset_tables: number
  }
  chat_transfer: {
    scanned: number
    parsed: number
    ignored: number
    reset_tables: number
    advanced_tables: number
  }
}

export interface FreebillSheetWorkbookSheet {
  key: string
  title: string
  sheet_id: number
  row_count: number
  updated_at: number
}

export interface FreebillSheetWorkbook {
  workbook: {
    id: number
    title: string
    updated_at: number
  }
  sheets: FreebillSheetWorkbookSheet[]
  refreshed_at: number
}

export async function fetchFreebillStatus() {
  const response = await api.get<FreebillStatus>('/freebill/status')
  return response.data
}

export async function fetchFreebillDashboard(params: FreebillQueryParams = {}) {
  const response = await api.get<FreebillDashboard>('/freebill/dashboard', { params })
  return response.data
}

export async function fetchFreebillDashboardByProgram(payload: FreebillDashboardProgramRequest) {
  const response = await api.post<FreebillDashboard>('/freebill/dashboard-program', payload)
  return response.data
}

export async function fetchFreebillRecords(params: FreebillQueryParams = {}) {
  const response = await api.get<FreebillRecordPage>('/freebill/records', { params })
  return response.data
}

export async function fetchFreebillCategoryBranchRecords(payload: FreebillCategoryBranchRecordsRequest) {
  const response = await api.post<FreebillRecordPage>('/freebill/category-branch-records', payload)
  return response.data
}

export async function fetchFreebillFilterOptions() {
  const response = await api.get<FreebillFilterOptions>('/freebill/filter-options')
  return response.data
}

export async function fetchFreebillInterpretRules() {
  const response = await api.get<FreebillInterpretRulesPayload>('/freebill/interpret-rules')
  return response.data
}

export async function saveFreebillInterpretRules(rules: FreebillInterpretRule[], settings?: FreebillInterpretRuleSettings) {
  const response = await api.put<FreebillInterpretRulesPayload>('/freebill/interpret-rules', { rules, settings })
  return response.data
}

export async function recomputeFreebillInterpretRules() {
  const response = await api.post<FreebillInterpretRecomputeResult>('/freebill/interpret-rules/recompute', {}, {
    timeout: 120000,
  })
  return response.data
}

export async function saveFreebillRecordManualOverrides(
  tradeNo: string,
  overrides: Record<string, unknown>,
  note?: string,
) {
  const response = await api.put('/freebill/record-manual-overrides', {
    trade_no: tradeNo,
    overrides,
    note,
  })
  return response.data as { trade_no: string; updated: number; overrides: Record<string, unknown> }
}

export async function saveFreebillCategoryBranchManualOverrides(payload: {
  program: FreebillProgramChannel
  programs?: FreebillProgramChannel[]
  path: FreebillCategoryPathItem[]
  overrides: Record<string, unknown>
  note?: string
}) {
  const response = await api.put('/freebill/category-branch-manual-overrides', payload)
  return response.data as { matched: number; updated: number; overrides: Record<string, unknown> }
}

export async function clearFreebillRecordOverrides(tradeNos: string[]) {
  const response = await api.post('/freebill/record-overrides/clear', { trade_nos: tradeNos })
  return response.data as { requested: number; cleared: number; missing_trade_nos: string[] }
}

export async function importFreebillFiles(source: FreebillImportSource, files: File[]) {
  const formData = new FormData()
  files.forEach((file) => formData.append('files', file))
  const response = await api.post<FreebillImportResult>(`/freebill/import/${source}`, formData, {
    headers: { 'Content-Type': 'multipart/form-data' },
    timeout: 120000,
  })
  return response.data
}

export async function syncFreebillFromLocalWechat() {
  const response = await api.post<FreebillWechatLocalSyncResult>('/freebill/sources/wechat-local/sync', {}, {
    timeout: 120000,
  })
  return response.data
}

export async function fetchFreebillSheetWorkbook() {
  const response = await api.get<FreebillSheetWorkbook | null>('/freebill/sheet-workbook')
  return response.data
}

export async function refreshFreebillSheetWorkbook() {
  const response = await api.post<FreebillSheetWorkbook>('/freebill/sheet-workbook/refresh', {}, {
    timeout: 120000,
  })
  return response.data
}

export function createFreebillProgramMatcher(kind: FreebillProgramMatcherKind = 'all'): FreebillProgramMatcher {
  return {
    kind,
    field: kind === 'field' ? 'create_time' : null,
    op: kind === 'field' ? 'between' : null,
    value: kind === 'full_text_contains' ? '' : undefined,
    values: [],
    ignore_case: true,
  }
}

export function createFreebillProgramRule(
  action: FreebillProgramRuleAction = 'include',
  kind: FreebillProgramMatcherKind = 'all',
): FreebillProgramRule {
  return {
    action,
    matcher: createFreebillProgramMatcher(kind),
  }
}

export function createFreebillIncludeAllProgram(): FreebillProgramChannel {
  return {
    default: false,
    rules: [
      createFreebillProgramRule('include', 'all'),
    ],
  }
}

export function normalizeFreebillProgramMatcher(value?: Partial<FreebillProgramMatcher> | null): FreebillProgramMatcher {
  const kind = value?.kind === 'none' || value?.kind === 'field' || value?.kind === 'full_text_contains'
    ? value.kind
    : 'all'
  const base = createFreebillProgramMatcher(kind)
  return {
    ...base,
    ...value,
    kind,
    field: typeof value?.field === 'string' ? value.field : base.field,
    op: value?.op ?? base.op,
    values: Array.isArray(value?.values) ? [...value.values] : [],
    ignore_case: typeof value?.ignore_case === 'boolean' ? value.ignore_case : true,
  }
}

export function normalizeFreebillProgramRule(value?: Partial<FreebillProgramRule> | null): FreebillProgramRule {
  return {
    action: value?.action === 'exclude' || value?.action === 'filter' ? value.action : 'include',
    matcher: normalizeFreebillProgramMatcher(value?.matcher),
  }
}

export function normalizeFreebillProgramChannel(value?: Partial<FreebillProgramChannel> | null): FreebillProgramChannel {
  return {
    default: typeof value?.default === 'boolean' ? value.default : false,
    rules: Array.isArray(value?.rules) ? value.rules.map((rule) => normalizeFreebillProgramRule(rule)) : [],
  }
}

export function cloneFreebillProgramChannel(value?: Partial<FreebillProgramChannel> | null): FreebillProgramChannel {
  return JSON.parse(JSON.stringify(normalizeFreebillProgramChannel(value))) as FreebillProgramChannel
}

export function createFreebillDateRangeRule(field: string, startDate: string, endDate: string): FreebillProgramRule {
  return {
    action: 'filter',
    matcher: {
      kind: 'field',
      field,
      op: 'between',
      values: [startDate, endDate],
      ignore_case: true,
    },
  }
}

export function upsertFreebillDateRangeRule(
  program: Partial<FreebillProgramChannel> | null | undefined,
  field: string,
  startDate: string,
  endDate: string,
): FreebillProgramChannel {
  const draft = cloneFreebillProgramChannel(program)
  const index = draft.rules.findIndex((rule) => (
    rule.matcher.kind === 'field'
    && rule.matcher.field === field
  ))
  const nextRule = createFreebillDateRangeRule(field, startDate, endDate)
  let keepIndex = index
  if (index >= 0) {
    draft.rules[index] = nextRule
  } else {
    draft.rules.push(nextRule)
    keepIndex = draft.rules.length - 1
  }
  draft.rules = draft.rules.filter((rule, ruleIndex) => (
    ruleIndex === keepIndex
    || rule.matcher.kind !== 'field'
    || rule.matcher.field !== field
  ))
  return normalizeFreebillProgramChannel(draft)
}

export function removeFreebillDateRangeRules(
  program: Partial<FreebillProgramChannel> | null | undefined,
  field: string,
): FreebillProgramChannel {
  const draft = cloneFreebillProgramChannel(program)
  draft.rules = draft.rules.filter((rule) => !(rule.matcher.kind === 'field' && rule.matcher.field === field))
  if (!draft.rules.length) {
    return createFreebillIncludeAllProgram()
  }
  return normalizeFreebillProgramChannel(draft)
}
